File: tiq-e/array/9/main.ts
import {performance} from 'perf_hooks';

function twoSum(nums: number[], target: number): number[] {
  const m: Map<number, number> = new Map();

  for (let i = 0; i < nums.length; i++) {
    if (m.has(nums[i])) {
      return [m.get(nums[i])!, i];
    }
    m.set(target - nums[i], i);
  }

  return [];
}

let arr = [2, 7, 11, 15];

let start = performance.now();
let result = twoSum(arr, 9);
let end = performance.now();
console.log('Duration:', end - start);
console.log('Result:', result)

File: tiq-e/array/9/main.rs
use std::collections::HashMap;
use std::time::Instant;

struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut m: HashMap<i32, i32> = HashMap::new();
        for (i, v) in nums.iter().enumerate() {
            match m.get(&(target - *v)) {
                Some(&i2) => return vec![i as i32, i2],
                None => m.insert(*v, i as i32),
            };
        }
        vec![]
    }
}

fn main() {
    let arr = Vec::from([2, 7, 11, 15]);

    let start = Instant::now();
    let result = Solution::two_sum(arr, 9);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
